[Executorch] Refactor op_mul's broadcasting utils

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #8257
* #8256
* #8255
* #8205
* __->__ #8204

Summary:
Refactoring broadcast handling utils that were added for op_mul. This is
in prepartion use these utils to handle broadcast for other ops such as
add, sub, div.
Plus remove a redundant test

Test Plan:
optimized_kernels_test in CI

Reviewers:

Subscribers:

Tasks:

Tags:

cc @larryliu0820 @manuelcandales

Differential Revision: [D69491816](https://our.internmc.facebook.com/intern/diff/D69491816)

kimishpatel committed:
[Executorch] Refactor op_mul's broadcasting utils

Summary:
Refactoring broadcast handling utils that were added for op_mul. This is
in prepartion use these utils to handle broadcast for other ops such as
add, sub, div.
Plus remove a redundant test

Test Plan:
optimized_kernels_test in CI

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Refactor op_mul's broadcasting utils"


Summary:
Refactoring broadcast handling utils that were added for op_mul. This is
in prepartion use these utils to handle broadcast for other ops such as
add, sub, div.
Plus remove a redundant test

Test Plan:
optimized_kernels_test in CI

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Refactor op_mul's broadcasting utils"


Summary:
Refactoring broadcast handling utils that were added for op_mul. This is
in prepartion use these utils to handle broadcast for other ops such as
add, sub, div.
Plus remove a redundant test

Test Plan:
optimized_kernels_test in CI

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

swolchok commented on kernels/optimized/cpu/op_mul.cpp:
```diff
@@ -238,7 +130,9 @@ Tensor& opt_mul_out(
     });
   } else if (selected_optimized_path != ElementwiseOptimizedPath::kNone) {
-    return handle_broadcast_mul(ctx, a, b, out, selected_optimized_path);
+    auto mul_lambda = [](auto x, auto y) { return x * y; };
+    return torch::executor::handle_broadcast_elementwise(
+        ctx, mul_lambda, a, b, out, selected_optimized_path);
```
nit: mul_lambda is [`std::multiplies`](https://en.cppreference.com/w/cpp/utility/functional/multiplies), up to you which is more readable though

swolchok reviewed (approved):
I sure wish GitHub flagged copied/moved code for easier review

swolchok commented on kernels/optimized/cpu/binary_ops.h:
```diff
@@ -190,5 +191,116 @@ std::array<int32_t, 3> inline get_normalized_tensor_size(
+      "Failed to resize output tensor.");
+  const size_t outer_size = getLeadingDims(out, out.dim() - 1);
+  const auto broadcast_size = out.size(out.dim() - 1);
+  ET_SWITCH_REALB_TYPES(out_type, ctx, "mul.out", CTYPE, [&]() {
```
this isn't in mul anymore, but I see it's fixed in the next PR, close enough

kimishpatel committed:
Update on "[Executorch] Refactor op_mul's broadcasting utils"


Summary:
Refactoring broadcast handling utils that were added for op_mul. This is
in prepartion use these utils to handle broadcast for other ops such as
add, sub, div.
Plus remove a redundant test

Test Plan:
optimized_kernels_test in CI

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Refactor op_mul's broadcasting utils"


Summary:
Refactoring broadcast handling utils that were added for op_mul. This is
in prepartion use these utils to handle broadcast for other ops such as
add, sub, div.
Plus remove a redundant test

Test Plan:
optimized_kernels_test in CI

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Refactor op_mul's broadcasting utils"


Summary:
Refactoring broadcast handling utils that were added for op_mul. This is
in prepartion use these utils to handle broadcast for other ops such as
add, sub, div.
Plus remove a redundant test

Test Plan:
optimized_kernels_test in CI

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Refactor op_mul's broadcasting utils"


Summary:
Refactoring broadcast handling utils that were added for op_mul. This is
in prepartion use these utils to handle broadcast for other ops such as
add, sub, div.
Plus remove a redundant test

Test Plan:
optimized_kernels_test in CI

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

swolchok commented on kernels/optimized/cpu/binary_ops.h:
```diff
@@ -190,5 +191,111 @@ std::array<int32_t, 3> inline get_normalized_tensor_size(
+  if ((selected_optimized_path ==
+       ElementwiseOptimizedPath::kBroadcastLastDim) ||
+      (selected_optimized_path ==
+       ElementwiseOptimizedPath::kBroadcastLastDimReverseArguments)) {
```
it occurs to me that we should separate the selected algorithm from whether to reverse arguments or not to make this read nicer, but that definitely doesn't have to go in this PR

swolchok reviewed (approved):
seems fine

kimishpatel committed:
Update on "[Executorch] Refactor op_mul's broadcasting utils"


Summary:
Refactoring broadcast handling utils that were added for op_mul. This is
in prepartion use these utils to handle broadcast for other ops such as
add, sub, div.
Plus remove a redundant test

Test Plan:
optimized_kernels_test in CI

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491816](https://our.internmc.facebook.com/intern/diff/D69491816)

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Refactor op_mul's broadcasting utils"


Summary:
Refactoring broadcast handling utils that were added for op_mul. This is
in prepartion use these utils to handle broadcast for other ops such as
add, sub, div.
Plus remove a redundant test

Test Plan:
optimized_kernels_test in CI

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491816](https://our.internmc.facebook.com/intern/diff/D69491816)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Refactor op_mul's broadcasting utils"


Summary:
Refactoring broadcast handling utils that were added for op_mul. This is
in prepartion use these utils to handle broadcast for other ops such as
add, sub, div.
Plus remove a redundant test

Test Plan:
optimized_kernels_test in CI

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491816](https://our.internmc.facebook.com/intern/diff/D69491816)

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Refactor op_mul's broadcasting utils"


Summary:
Refactoring broadcast handling utils that were added for op_mul. This is
in prepartion use these utils to handle broadcast for other ops such as
add, sub, div.
Plus remove a redundant test

Test Plan:
optimized_kernels_test in CI

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491816](https://our.internmc.facebook.com/intern/diff/D69491816)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69491816).

kimishpatel committed:
Update on "[Executorch] Refactor op_mul's broadcasting utils"


Summary:
Refactoring broadcast handling utils that were added for op_mul. This is
in prepartion use these utils to handle broadcast for other ops such as
add, sub, div.
Plus remove a redundant test

Test Plan:
optimized_kernels_test in CI

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491816](https://our.internmc.facebook.com/intern/diff/D69491816)

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Refactor op_mul's broadcasting utils"


Summary:
Refactoring broadcast handling utils that were added for op_mul. This is
in prepartion use these utils to handle broadcast for other ops such as
add, sub, div.
Plus remove a redundant test

Test Plan:
optimized_kernels_test in CI

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491816](https://our.internmc.facebook.com/intern/diff/D69491816)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69491816).
